book/basics/units_basics.nim

```diff
@@ -154,6 +154,37 @@ nbCodeInBlock:
   let textUnit = 1.`N*s`
   echo unicodeUnit == textUnit
 
+nbText: """
+### Arithmetics with units
+Units can of course be used in formulas. The supported operators are `+`, `-`, `*`, `/` and `^` (Note: `^` only work with integer exponents!). 
+You don't have to `defUnit` the resulting unit of an operation, it's done automatically.
+There are a few rules worth remembering, some more obvious than others:
+
+- You can only add and subtract units of the same quantity (hopefully pretty obvious). 
+Example: `1.kg + 1.lbs` works. `1.kg + 1.m` doesn't work.
+- You can multiply and divide units of different quantities. Example: `1.kg * 1.m` works.
+- If it is ambiguious what the resulting unit should be, they are converted to base SI units. Example: `1.kg + 1.lbs` will result in `1.45359 KiloGram`.
+- Conversion from prefixed to non-prefixed units only happen if multiple different units or prefixes of the same quantity are mixed.
+Example: `1.mm * 1.kg` results in `1 KiloGram•MilliMeter` while `1.mm * 1.cm` results in `1e-005 Meter²`.
+- Division by same quantities gives a `UnitLess` result. Example: `1.cm / 1.m` gives `0.01 UnitLess`.
+
+Let's see some actual code now! You are probably familiar with the harmonic oscillator:
+$$y(t) = A \cos (ω t + φ)$$
+where $ω$ is the frequency in $rad ⋅ s^{-1}$, $φ$ is an angle and $A$ is the maximum amplitude in $cm$.
+Let's get coding now!
+"""
+
+nbCode: import math
+nbCodeInBlock:
+  proc y(t: Second): CentiMeter =
+    let ω = 100.rad•s⁻¹ ## or 100.`rad*s^-1`
+    let A = 10.cm
+    let φ = Pi.rad
+    let argument = ω * t + φ
+    result = A * cos(argument.to(UnitLess))
```

Comment:
The explicit conversion to `UnitLess` shouldn't be necessary, because of the converter. If you want to do it to be explicit, I'd add a comment.

Reply:
Well, it is necessary atm. I think it has to do with the fact `argument` has type `Meter⁻¹•Meter` while radians are defined as the reverse `Meter•Meter⁻¹`

Reply:
Ouch, I suppose we need a smarter converter. :rofl: Not sure if a converter can be generic. If so I can write a `IsUnitLess` concept that allows to match these things. Thanks for the detective work. :smile:

Reply:
First attempt at that causes a compiler segfault, yay.

Reply:
Hehe. "detective job" == "just write code and hope Vindaar spots it" ? 🤣 On the topic of this, wouldn't it make sense to store compound units in a non-ordered manner? In a set or table instead of a seq for example? And surely there must be better ways of expressing unitless units like angles than with `m/m`? `distinct UnitLess` doesn't work?😅

Reply:
> In a set or table instead of a seq for example?

That's exactly what I do (as a table). 

The problem is we need to express the units as a name to write them out of course. Currently there is *some* ordering done, but exponents apparently are not taken into account. Therefore we can end up with `Meter•Meter⁻¹` and `Meter⁻¹•Meter`. That should be an easy fix though. The concept would allow for a more general solution though.

> And surely there must be better ways of expressing unitless units like angles than with m/m? distinct UnitLess doesn't work?

Well, yes and no. `distinct UnitLess` works of course. But the correct unit of radians is plain and simply `Meter•Meter⁻¹`. So I'd like to keep that.

Reply:
Oops 😅 must have been me thinking I remembered the Unchained code again. Just have a vague memory of a sew of compound units somewhere 🙈

Having one canonical representation of a unit would be nice and that sounds like a step in the right direction 😄

> Well, yes and no. distinct UnitLess works of course. But the correct unit of radians is plain and simply Meter•Meter⁻¹. So I'd like to keep that.

You must be kidding me! 🤯 How do we differentiate it from say 1.m / 1.m? Or is it only possible to get a Meter•Meter⁻¹ from defining a variable as it?

Reply:
> You must be kidding me! exploding_head How do we differentiate it from say 1.m / 1.m? Or is it only possible to get a Meter•Meter⁻¹ from defining a variable as it?

Well. We don't in a sense. That's why these should be convertible to `UnitLess`. And that's also why even a perfect unit system is technically incomplete due to different quantities that have the same unit. For all intents and purposes though they are the same anyway.

But yes, in *principle* maths acting on 'unity' numbers should produce `UnitLess` instead of `Meter•Meter⁻¹`. That just doesn't happen here as the units we're dealing with are Radians instead of actual meter / meter things. There just appears to be an issue that under certain contexts the information of `Radian` as a distinct unit gets lost in maths of `ω * t` for instance. That should result in pure `Radian` as a unit of course.

Radians were added kind of quickly by me, so it's expected that there's more edge cases I didn't consider.

Reply:
> Oops sweat_smile must have been me thinking I remembered the Unchained code again. Just have a vague memory of a sew of compound units somewhere see_no_evil

Oh, and well. That's not your fault. The internal storage of units is done in a sequence that stores `CTUnit` things. 

The `Table` I mention is only used to determine if two units are the same quantity here:

https://github.com/SciNim/Unchained/blob/master/src/unchained/units.nim#L300-L307

(in `commonQuantity` via `toCTQuantity`). Then there's also usage of sets here:

https://github.com/SciNim/Unchained/blob/master/src/unchained/units.nim#L827-L859

in `simplify` to in principle turn `Meter•Meter⁻¹` to `UnitLess`.

In particular the latter approach needs to be cleaned up though. Maybe I can combine it with the `Table` in some way or maybe not. Some of this logic is in need of a refactor for sure.

Reply:
> But yes, in principle maths acting on 'unity' numbers should produce UnitLess instead of Meter•Meter⁻¹. That just doesn't happen here as the units we're dealing with are Radians instead of actual meter / meter things. There just appears to be an issue that under certain contexts the information of Radian as a distinct unit gets lost in maths of ω * t for instance. That should result in pure Radian as a unit of course.
>
>Radians were added kind of quickly by me, so it's expected that there's more edge cases I didn't consider.

OK 😄 There are so many units so it's understandable that not all of them are working flawlessly. Do you think this is something you will be able to fix before we publish this post? Otherwise I'll just add a note that it's a bug atm :)

> Oh, and well. That's not your fault. The internal storage of units is done in a sequence that stores CTUnit things.
> The Table I mention is only used to determine if two units are the same quantity here:

Hehe, look at that :o Seems like you are using tables and sets where they are important :D One can *always* refactor one's code to make it better 🙃

Reply:
>  Do you think this is something you will be able to fix before we publish this post? Otherwise I'll just add a note that it's a bug atm :)

Depends on when you want to have it published. I can certainly take an hour or two tomorrow to try to fix it.

Reply:
> Depends on when you want to have it published. I can certainly take an hour or two tomorrow to try to fix it.

I'll be away from my computer all next week so I'm not in any hurry at all, take your time :) It's exactly as if there is anyone out there refreshing the getting-started page waiting for this 🤣